### PR TITLE
Adds a "Joseki Explorer" React App

### PR DESCRIPTION
**Pull request by amj**
Here's something: an interactive joseki explorer built with yet-another-tech stack.  At least the backend is a basic flask microservice...

Anyway, this collects the opening frequency work into one directory (oneoffs/joseki) and should cover most of the stuff.

### CODE REVIEW

File: oneoffs/joseki/joseki_query.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+    d = json.loads(flask.request.data.decode('utf-8'))
+    prefix = d['params']['prefix']
+    run = d['params']['run']
+
+    db = get_db()
+
+    res = {'count': 0, 'next_moves': {}}
+
+    if not prefix:
+        nexts = db.execute("select distinct(substr(seq, 0, 7)), count(*) from joseki group  by 1;").fetchall()
```
Comment (sethtroisi) [inline oneoffs/joseki/joseki_query.py]: comment describing this

File: oneoffs/joseki/joseki_query.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+                               (s_id, run)).fetchone()[0]
+
+    if not nexts:
+        print("No next moves found, post params:", d['params'])
+        return flask.Response(json.dumps(res), mimetype='text/json')
+
+    next_moves = {}
+    tot = 0
+    for nxt, ct in nexts:
+        next_moves[nxt] = ct
```
Comment (sethtroisi) [inline oneoffs/joseki/joseki_query.py]: next_moves = dict(nexts)
tot = sum(next_moves.values())

File: oneoffs/joseki/joseki_query.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+
+    res = {'count': 0, 'next_moves': {}}
+
+    if not prefix:
+        nexts = db.execute("select distinct(substr(seq, 0, 7)), count(*) from joseki group  by 1;").fetchall()
+        total = sum([n[1] for n in nexts])
+    else:
+        s_id = db.execute("select id from joseki where seq=?", (prefix,)).fetchone()
+        if not s_id:
+            return flask.Response(json.dumps(res), mimetype='text/json')
```
Comment (sethtroisi) [inline oneoffs/joseki/joseki_query.py]: consider:

from flask import jsonify
return jsonify(res)

this handles setting status=200 and mimetype

File: oneoffs/joseki/joseki_query.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+            start = timer()
+            nexts = db.execute("""
+                               select next_move, sum(nm.count)
+                               from next_moves as nm join joseki_counts as jc
+                               on jc.id == nm.joseki_hour_id
+                               where nm.seq_id = ? and jc.run = ? group by 1
+                               """, (s_id, run)).fetchall()
+            end = timer()
+            print('%.4f seconds for fancy join.' % (end-start,))
+            total = db.execute("select sum(count) from joseki_counts where seq_id = ? and run=?",
```
Comment (sethtroisi) [inline oneoffs/joseki/joseki_query.py]: be consistent A = ? or A=?

File: oneoffs/joseki/joseki_query.py [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+def games():
+    d = json.loads(flask.request.data.decode('utf-8'))
+    prefix = d['params']['sgf']
+    sort_hour = d['params']['sort']
+    run = d['params']['run']
+    # "page" is 1-indexed, so subtract 1 to get the proper OFFSET.
+    page = d['params']['page'] - 1 
+    db = get_db()
+
+    if (sort_hour.lower() != 'desc' and sort_hour.lower() != 'asc'):
```
Comment (sethtroisi) [inline oneoffs/joseki/joseki_query.py]: sort_hour.lower() not in ('desc', 'asc')

File: oneoffs/joseki/joseki_query.py [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+        print("Invalid input for sort_hour param: ", sort_hour)
+        return flask.Response(json.dumps({'rows': []}), mimetype='text/json')
+
+    s_id = db.execute("select id from joseki where seq=?", (prefix,)).fetchone()
```
Comment (sethtroisi) [inline oneoffs/joseki/joseki_query.py]: this is repeated several times consider extracting to `get_seq_id(seq)`

File: oneoffs/joseki/joseki_query.py [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+        rows = db.execute(q, (s_id, page * 30)).fetchall()
+    res = [ {'game': os.path.basename(r[0]), 'hour': r[1],
+             'run': r[2], 'winrate': r[3]} for r in rows]
+
+    return flask.Response(json.dumps({'rows': res}), mimetype='text/json')
+
+
+@app.route("/search", methods=["POST"])
+def search():
+    d = json.loads(flask.request.data.decode('utf-8'))
```
Comment (sethtroisi) [inline oneoffs/joseki/joseki_query.py]: if you set mimetype in the post request, then you can simply this to request.get_json()

https://stackoverflow.com/a/20001283/459714

File: oneoffs/joseki/joseki_query.py [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+    ts = lambda hr: int(dt.datetime.strptime(hr, "%Y-%m-%d-%H").timestamp())
+    ranges = openings.run_time_ranges(get_db())
+    interps = openings.build_run_time_transformers(ranges)
+
+    runs = sorted(ranges.keys())
+
+    cols = []
+    cols.append({'id': 'time', 'label': '% of Training', 'type': 'number'})
+    for run in runs:
+        cols.append({'id': run + 'count', 'label': run + ' times seen', 'type': 'number'})
```
Comment (sethtroisi) [inline oneoffs/joseki/joseki_query.py]: run + 'count' does this need a space?

File: oneoffs/joseki/joseki_query.py [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+
+    for r, hr, ct in sequence_counts:
+        key = interps[r](ts(hr))
+        idx = runs.index(r)
+        rows[key][idx] = ct
+
+    row_data = [ {'c': [ {'v': key} ] + [{'v': v if v else None} for v in value ] }
+                for key,value in rows.items()]
+    obj = {'cols': cols, "rows": row_data, "sequence": query}
+    data.append(obj)
```
Comment (sethtroisi) [inline oneoffs/joseki/joseki_query.py]: what is the purpose of this line?

File: oneoffs/joseki/opening_freqs.py [hunk 10]
```diff
@@ -33,9 +49,9 @@
 flags.DEFINE_string("db_path", 'joseki.db', "Path to josekidb")
 flags.DEFINE_string("run_name", '',
                     "If the games should be associated with a 'run' key (e.g., 'v17')")
-flags.DEFINE_integer("max_patterns", 5000,
+flags.DEFINE_integer("most_common", 3000,
                      "Number of most-common-joseki to save per hour")
-flags.DEFINE_integer("threads", 8, "Number of threads to use.")
```
Comment (sethtroisi) [inline oneoffs/joseki/opening_freqs.py]: I'd just drop use the default of mp.pool

File: oneoffs/joseki/opening_freqs.py [hunk 11]
```diff
@@ -47,31 +63,87 @@ def sha(sequence):
...
+        key = (t-min_) * (1/(max_-min_))
+        return "%0.3f" % (int(buckets*key) / (buckets / 100.0))
+
+    for run, range_ in ranges.items():
+        funcs[run] = functools.partial(f, min_=range_[0], max_=range_[1])
+
+    return funcs
+
+
+def namedtuple_factory(cursor, row):
```
Comment (sethtroisi) [inline oneoffs/joseki/opening_freqs.py]: unused, I also think that sqlite3.Row does some of this.

File: oneoffs/joseki/opening_freqs.py [hunk 12]
```diff
@@ -157,12 +240,13 @@ def extract_corners(game_path):
                 canonical = False
             elif canonical is False:
                 x, y = y, x
```
Comment (sethtroisi) [inline oneoffs/joseki/opening_freqs.py]: maybe extract to "canonical corner"

File: oneoffs/joseki/opening_freqs.py [hunk 13]
```diff
@@ -157,12 +240,13 @@ def extract_corners(game_path):
...
-            seq += sgf_format((color, (18 - y, x)))
-
-            if idx > MIN_JOSEKI_LENGTH:
-                sequence_counts[seq] += 1
+            next_move = sgf_format((color, (18 - y, x)))
+            if seq:
+                next_moves[seq][next_move] += 1
```
Comment (sethtroisi) [inline oneoffs/joseki/opening_freqs.py]: I'd just add the empty sequence here and be damned what it does

File: oneoffs/joseki/opening_freqs.py [hunk 14]
```diff
@@ -188,26 +274,40 @@ def analyze_dir(directory):
...
-            cur.execute(""" INSERT INTO joseki(seq, length, num_tenukis) VALUES(?, ?, ?)
-                 ON CONFLICT DO NOTHING """,
-                        (c[0], c[0].find(';'), count_tenukis(c[0])))
+
+            s_id = cur.execute('select id from joseki where seq=?', (seq,)).fetchone()
+            if s_id:
+                s_id = s_id[0]
+            else:
+                cur.execute(""" INSERT INTO joseki(seq, length, num_tenukis) VALUES(?, ?, ?) """,
+                           (seq, len(seq.split(';')[:-1]), count_tenukis(seq)))
```
Comment (sethtroisi) [inline oneoffs/joseki/opening_freqs.py]: seq.count(';') - 1

File: oneoffs/joseki/opening_freqs.py [hunk 15]
```diff
@@ -188,26 +274,40 @@ def analyze_dir(directory):
...
                  """,
-                        (c[0], hr, c[1], FLAGS.run_name, example_sgfs[c[0]], c[1]))
-            cur.execute("""
-                  INSERT INTO game_counts_by_hour(hour, count) VALUES (?, ?) ON CONFLICT DO NOTHING 
-                        """,
-                        (hr, c[1]))
+                        (s_id, hr, count, FLAGS.run_name, b_wins[seq], example_sgfs[seq]))
+            jc_id = cur.lastrowid
+
+            for n,ct in next_moves[seq].items():
```
Comment (sethtroisi) [inline oneoffs/joseki/opening_freqs.py]: next_move, next_count to be more precise

File: oneoffs/joseki/public/robots.txt [hunk 16]
```diff
@@ -0,0 +1,2 @@
+# https://www.robotstxt.org/robotstxt.html
+User-agent: *
```
Comment (sethtroisi) [inline oneoffs/joseki/public/robots.txt]: This is pretty boring, I'd just delete it

File: oneoffs/joseki/joseki_query.py [hunk 17]
```diff
@@ -0,0 +1,194 @@
...
+import flask
+
+import oneoffs.joseki.opening_freqs as openings
+
+# static_folder is location of npm build
+app = Flask(__name__, static_url_path="", static_folder="./build")
+
+# Path to database relative to joseki_query.py, with pre-extracted joseki
+# information (see 'opening_frequencies.py')
+DATABASE = '/data/sgf/wr_joseki_3000.db'
```
Comment (sethtroisi) [inline oneoffs/joseki/joseki_query.py]: this shouldn't be hardcoded

File: oneoffs/joseki/joseki_query.py [hunk 18]
```diff
@@ -0,0 +1,194 @@
...
+def get_sequence_hour_counts(seq):
+    cur = get_db().execute('''
+                     select run, hour, count from joseki_counts
+                           where seq_id in (select id from joseki where seq = ?);
+                     ''', (seq, ))
+    return list(cur.fetchall())
+
+def seq_id_or_none(db, seq):
+    s_id = db.execute("select id from joseki where seq = ?", (seq,)).fetchone()
+    if not s_id:
```
Comment (sethtroisi) [inline oneoffs/joseki/joseki_query.py]: consider `return s_id[0] if s_id else None`

File: oneoffs/joseki/joseki_query.py [hunk 19]
```diff
@@ -0,0 +1,194 @@
...
+        return None
+    else:
+        return s_id[0]
+
+@app.route("/")
+def index():
+    return app.send_static_file("index.html")
+
+@app.route("/nexts", methods=["POST"])
+def nexts():
```
Comment (sethtroisi) [inline oneoffs/joseki/joseki_query.py]: consider: next_stone, next_moves, next_play, or next action

File: oneoffs/joseki/opening_freqs.py [hunk 20]
```diff
@@ -122,8 +186,19 @@ def extract_corners(game_path):
         _, moves = sgf_moves.get_setup_and_moves(g)
     except BaseException:
         print("bad file: ", game_path)
-        return Counter()
+        return Counter(), {}
+
+    return (extract_corners(moves), 1 if g.get_winner() == 'b' else 0)
```
Comment (sethtroisi) [inline oneoffs/joseki/opening_freqs.py]: what is get_winner() from?
should we .lower() just to be safe?

File: oneoffs/joseki/public/index.html [hunk 21]
```diff
@@ -0,0 +1,43 @@
+<!DOCTYPE html>
```
Comment (sethtroisi) [inline oneoffs/joseki/public/index.html]: I can't leave a comment on favicon.ico, if it's the default react favicon please remove it

File: oneoffs/joseki/src/App.js [hunk 22]
```diff
@@ -0,0 +1,455 @@
...
+      moves: "",
+      highlights: null,
+      other_highlights: null,
+      chartData: null,
+      tableData: null,
+      tablePage: 1,
+      search_enabled: false,
+      run: null,
+      doneMessage: false,
+    }, () => {
```
Comment (sethtroisi) [inline oneoffs/joseki/src/App.js]: is the "() => { this.updateHeatmap() }" a callback on success of setState?

File: oneoffs/joseki/opening_freqs.py (hunk 20 above)
Comment (amj) [reply oneoffs/joseki/opening_freqs.py]: it's from sgfmill, it seems to call lower on it already.

File: oneoffs/joseki/public/index.html (hunk 21 above)
Comment (amj) [reply oneoffs/joseki/public/index.html]: nope, i made a new one :)

File: oneoffs/joseki/src/App.js (hunk 22 above)
Comment (amj) [reply oneoffs/joseki/src/App.js]: yes.

File: oneoffs/joseki/joseki_query.py (hunk 17 above)
Comment (amj) [reply oneoffs/joseki/joseki_query.py]: Done.

File: oneoffs/joseki/joseki_query.py (hunk 18 above)
Comment (amj) [reply oneoffs/joseki/joseki_query.py]: Done.

File: oneoffs/joseki/joseki_query.py (hunk 19 above)
Comment (amj) [reply oneoffs/joseki/joseki_query.py]: python changed only.  Done.

File: oneoffs/joseki/opening_freqs.py (hunk 20 above)
Comment (amj) [reply oneoffs/joseki/opening_freqs.py]: Done.

**Pull request merged**